Simplenote/src/main/java/com/automattic/simplenote/TagsActivity.java

```diff
@@ -82,48 +119,96 @@ public void onCreate(Bundle savedInstanceState) {
-                    intent.putExtra(MorphSetup.EXTRA_SHARED_ELEMENT_COLOR_END, ThemeUtils.getColorFromAttribute(TagsActivity.this, R.attr.drawerBackgroundColor));
-                    intent.putExtra(MorphSetup.EXTRA_SHARED_ELEMENT_COLOR_START, ThemeUtils.getColorFromAttribute(TagsActivity.this, R.attr.fabColor));
-                    ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(TagsActivity.this, mButtonAdd, "shared_button");
-                    startActivityForResult(intent, REQUEST_ADD_TAG, options.toBundle());
-                }
-            }
-        );
-        mButtonAdd.setOnLongClickListener(
-            new View.OnLongClickListener() {
-                @Override
-                public boolean onLongClick(View v) {
-                    if (v.isHapticFeedbackEnabled()) {
-                        v.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
+        mButtonAdd.setOnClickListener(v -> viewModel.clickAddTag());
+        mButtonAdd.setOnLongClickListener(v -> {
+            viewModel.longClickAddTag();
+            return true;
+        });
+    }
+
+    private void setObservers() {
+        viewModel.getUiState().observe(this, uiState ->
+                tagItemAdapter.submitList(uiState.getTagItems(), () -> {
+                    if (uiState.getSearchUpdate()) {
+                        mTagsList.scrollToPosition(0);
+                        boolean isSearching = uiState.getSearchQuery() != null;
+                        checkEmptyList(isSearching);
                     }
+                })
+        );
 
-                    Toast.makeText(TagsActivity.this, getString(R.string.add_tag), Toast.LENGTH_SHORT).show();
-                    return true;
-                }
+        // Observe different events such as clicks on add tags, edit tags and delete tags
+        viewModel.getEvent().observe(this, event -> {
+            if (event instanceof TagsEvent.AddTagEvent) {
+                startAddTagActivity();
+            } else if (event instanceof TagsEvent.LongAddTagEvent) {
+                showLongAddToast();
+            } else if (event instanceof TagsEvent.FinishEvent) {
+                finish();
+            } else if (event instanceof TagsEvent.EditTagEvent) {
+                showTagDialogFragment((TagsEvent.EditTagEvent) event);
+            } else if (event instanceof TagsEvent.DeleteTagEvent) {
+                showDeleteDialog((TagsEvent.DeleteTagEvent) event);
+            } else if (event instanceof TagsEvent.LongDeleteTagEvent) {
+                showLongDeleteToast((TagsEvent.LongDeleteTagEvent) event);
             }
+        });
+    }
+
+    private void showLongAddToast() {
+        if (mButtonAdd.isHapticFeedbackEnabled()) {
+            mButtonAdd.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
+        }
+
+        Toast.makeText(TagsActivity.this, getString(R.string.add_tag), Toast.LENGTH_SHORT).show();
```

Comment:
in the past I found it helpful to add a class method for toasting. the toasts were all similar enough that they mostly just required passing the string. possibly the length.

```java
private void showLongAddToast() {
    if (mButtonAdd…) {
        …
    }

    toast(R.string.add_tag);
}

private void toast(@StringRes int resId) {
    toast(resId, Toast.LENGTH_SHORT);
}

private void toast(@StringReg int resId, @Duration int duration) {
    Toast.makeText(TagsActivity.this, resId, duration).show();
}
```

Reply:
This is a good idea. I have added just one method since both toasts have the same duration, but these are functions that can be added as utility functions for all the project. Fixed in fc1bf536bd26c1eee274a2c11c2290bfda37568a.

```kotlin
 private void toast(@StringRes int resId) {
        Toast.makeText(TagsActivity.this, getString(resId), Toast.LENGTH_SHORT).show();
 }
```